File: lib/core/text_pluralization.dart
/// French pluralization for noun phrases that may include adjectives.
///
/// Splits the input on spaces and pluralizes each token following common
/// French rules:
///   - words ending in s/x/z are invariable
///   - -al → -aux (Cheval → Chevaux)
///   - -au, -eau, -eu → +x (Veau → Veaux, Cheveu → Cheveux)
///   - everything else → +s
///
/// Returns [word] unchanged when [count] is 0 or 1.
///
/// Known limitations (uncommon for animal-category names; users can
/// rename if hit):
///   - prepositions inside compounds (e.g. "Cheval de trait") are
///     incorrectly pluralized
///   - the 7 nouns in -ou that take -oux (bijou, caillou, chou, genou,
///     hibou, joujou, pou) are pluralized as +s
///   - -al exceptions (bal, carnaval, festival, …) are pluralized -aux
String pluralizeFr(String word, int count) {
  if (count <= 1 || word.isEmpty) return word;
  return word.split(' ').map(_pluralizeWord).join(' ');
}

String _pluralizeWord(String w) {
  if (w.isEmpty) return w;
  final lower = w.toLowerCase();
  if (lower.endsWith('s') || lower.endsWith('x') || lower.endsWith('z')) {
    return w;
  }
  if (lower.endsWith('al')) {
    return '${w.substring(0, w.length - 2)}aux';
  }
  if (lower.endsWith('eau') || lower.endsWith('au') || lower.endsWith('eu')) {
    return '${w}x';
  }
  return '${w}s';
}

File: src/lib/text-pluralization.ts
export function pluralize(count: number, singular: string, plural: string): string {
  return Math.abs(count) <= 1 ? singular : plural;
}

/**
 * Returns the noun in the form matching `count` (singular for ≤1, else plural).
 * Accepts the input in either form: pluralizes a singular when needed,
 * singularizes a plural when needed.
 *
 * Plural rules: -s/-x/-z already plural; -al → -aux; -au/-eau/-eu → +x; else +s.
 * Singular rules: -eaux → -eau; -aux → -al; trailing -s → strip; -x/-z → as-is.
 *
 * Known limitations: -ou words taking -oux (bijou, etc.), -al exceptions
 * (bal, festival, …), and prepositions in compounds are not handled.
 */
export function pluralizeFr(word: string, count: number): string {
  if (word.length === 0) return word;
  if (count <= 1) return word.split(' ').map(singularizeFrWord).join(' ');
  return word.split(' ').map(pluralizeFrWord).join(' ');
}

function pluralizeFrWord(w: string): string {
  if (w.length === 0) return w;
  const lower = w.toLowerCase();
  if (lower.endsWith('s') || lower.endsWith('x') || lower.endsWith('z')) return w;
  if (lower.endsWith('al')) return `${w.slice(0, -2)}aux`;
  if (lower.endsWith('eau') || lower.endsWith('au') || lower.endsWith('eu')) return `${w}x`;
  return `${w}s`;
}

function singularizeFrWord(w: string): string {
  if (w.length === 0) return w;
  const lower = w.toLowerCase();
  if (lower.endsWith('eaux')) return w.slice(0, -1); // Veaux → Veau
  if (lower.endsWith('aux') && w.length > 3) return `${w.slice(0, -3)}al`; // Chevaux → Cheval
  if (lower.endsWith('s')) return w.slice(0, -1); // Moutons → Mouton
  return w; // already singular or unknown form
}
